### ISSUE TITLE
NoiseTorch unable to find valid pulse runtime path

### ISSUE THREAD

**Issue by SpecialBomb** [lawl/NoiseTorch#212]
I run `pulseaudio` as a user-level service upon launching `xinit` after tty login, the process is not system-wide or has escalated privileges.

Upon attempting to execute `noisetorch` as the README explained, it spits out this error (with `-log`):

```
2022/01/05 20:24:14 Application starting. Version: 0.11.4 (official)
2022/01/05 20:24:14 CAP_SYS_RESOURCE: true
2022/01/05 20:24:14 Checking if config needs to be initialized
2022/01/05 20:24:14 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
2022/01/05 20:24:14 Wrote temp librnnoise to: /tmp/librnnoise-322563810.so
2022/01/05 20:24:14 Couldn't resolve $XDG_CONFIG_HOME falling back to '/home/user/.config'
Couldn't create pulseaudio client: No valid directory for Pulse RuntimePath found
2022/01/05 20:24:14 Deleted temp librnnoise: /tmp/librnnoise-322563810.so
```

I could not find any documentation for this error, and also tried addressing some of the other problems.

Setting `XDG_CONFIG_HOME` resolves the falling back issue, however does not change the result.

I also attempted to set `PULSE_RUNTIME_PATH` to the current pulse runtime path (ex., `/home/user/.config/pulse/xxxxx-runtime`), which led to no change in output.

I tried to search for the error message in the repository to see what `noisetorch` was actually trying to find when executed, but I couldn't dissect the resulting code.

One thing I noticed is that `noisetorch` has a package available on Void Linux, the distro I am using. I have tried both the package and manually installing, both resulting in the same issue. The fact that the package is available leads me to think that there may be something different about my system that isn't standard to other Void distros.

My system:
Void Linux (Linux 5.13.19_1 x86_64)
Pulseaudio 15.0
NoiseTorch 0.11.4

**Comment by lawl** [comment]
Hey, as long as the fallback path for `XDG_CONFIG_HOME` is correct, its not an issue and you can ignore it, just debug output.

The problem about not finding the runtime path is real. The corresponding code is here: 
https://github.com/lawl/pulseaudio/blob/master/client.go#L347

Do you know where your pulse native socket is located?

**Comment by SpecialBomb** [comment]
>Hey, as long as the fallback path for XDG_CONFIG_HOME is correct, its not an issue and you can ignore it, just debug output.

Thought so, but it's always worth a try anyway.

>Do you know where your pulse native socket is located?

Yes, its located within my `/tmp/` directory. There is a symbolic link within my `/home/user/.config/pulse` directory that leads to the same runtime directory. `PULSE_RUNTIME_PATH` also leads to this directory, so I'm not sure why it wouldn't be working.

**Comment by lawl** [comment]
I suggest `strace`ing noisetorch to try and figure out why it can't find it.

**Comment by SpecialBomb** [comment]
What exactly should I be looking for?

**Comment by lawl** [comment]
If I run `strace noisetorch -l` on my machine, and scroll up a bit, i see the following.

```
newfstatat(AT_FDCWD, "/run/user/1000", {st_mode=S_IFDIR|0700, st_size=400, ...}, 0) = 0
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 7
connect(7, {sa_family=AF_UNIX, sun_path="/run/user/1000/pulse/native"}, 30) = 0
epoll_ctl(4, EPOLL_CTL_ADD, 7, {events=EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, data={u32=2455424648, u64=140447886003848}}) = 0
getsockname(7, {sa_family=AF_UNIX}, [112 => 2]) = 0
getpeername(7, {sa_family=AF_UNIX, sun_path="/run/user/1000/pulse/native"}, [112 => 30]) = 0
```

If you correlate that with the code I linked above, you can see, I have no `PULSE_RUNTIME_PATH` or `XDG_RUNTIME_DIR` set, so the code checks if `/run/user/1000` exists (it does) and then tries to connect to the expected path of the pulse socket `/run/user/1000/pulse/native`, which succeeds.

Now, let me set ma wrong `PULSE_RUNTIME_PATH` on purpose, so you can see the failure case.

```
mkdir /tmp/foobar
PULSE_RUNTIME_PATH=/tmp/foobar strace noisetorch -l
```

gives me

```
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 7
connect(7, {sa_family=AF_UNIX, sun_path="/tmp/foobar/native"}, 21) = -1 ENOENT (No such file or directory)
close(7)                                = 0
write(2, "Couldn't create pulseaudio clien"..., 100Couldn't create pulseaudio client: dial unix /tmp/foobar/native: connect: no such file or directory
) = 100
```

As you can see it directly tries to connect to `/tmp/foobar/native`, which does not exist. So I can look into `/tmp/foobar` and see, that it's indeed the empty directory we just created, and the return code `-1 ENOENT (No such file or directory)` is indeed correct, because we specified the wrong `PULSE_RUNTIME_PATH` .

Hope that helps you debug this issue, please let me know if/how you solved it and if this ends up being a quirk on your system or a bug in noisetorch.

**Comment by lawl** [comment]
Ping

**Comment by SpecialBomb** [comment]
Hey sorry! Just been caught up in life, I'll try to diagnose this tomorrow.

**Comment by SpecialBomb** [comment]
So, I got to it and finally got things working. I set my runtime path as I usually do, and it just decided to work. What's confusing is why it didn't work last time, since I did the same envar export to what I was pretty sure my pulse runtime directory. Because of this, I won't exactly say this issue is closed, but ill make sure to update whether it happens again in case if it's a weird edge case.
